aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,858 @@
+        // in the table. If one is absent, committing the transaction would leave Janus unstartable
+        // so we should rollback.
+        let current_keypairs = tx.get_global_hpke_keypairs().await?;
+        debug!(?current_keypairs, "table state after running key rotator");
+        if !current_keypairs
+            .iter()
+            .any(|keypair| keypair.state() == &HpkeKeyState::Active)
+        {
+            Err(DatastoreError::User(
+                anyhow!("unexpected state: no keypairs are active").into(),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    #[tracing::instrument(err, skip(tx, current_keypairs, available_ids))]
+    async fn run_hpke_for_ciphersuite(
+        tx: &Transaction<'_, C>,
+        config: &HpkeKeyRotatorConfig,
+        current_keypairs: &[GlobalHpkeKeypair],
+        available_ids: &mut impl Iterator<Item = HpkeConfigId>,
+        ciphersuite: &HpkeCiphersuite,
+    ) -> Result<(), DatastoreError> {
+        let clock = tx.clock();
+
+        let (pending_keypairs, active_keypairs) =
+            Self::partition_keypairs(current_keypairs, &ciphersuite);
+
+        let mut next_key = || {
+            generate_hpke_config_and_private_key(
+                available_ids.next().ok_or_else(|| {
+                    DatastoreError::User(anyhow!("global HPKE key ID space exhausted").into())
+                })?,
+                ciphersuite.kem_id(),
+                ciphersuite.kdf_id(),
+                ciphersuite.aead_id(),
+            )
+            .map_err(|e| DatastoreError::User(e.into()))
+        };
+
+        if active_keypairs.is_empty() && pending_keypairs.is_empty() {
+            // Bootstrapping case: there are no keys at all for this ciphersuite, so we need to
+            // insert one.
+            let new_keypair = next_key()?;
+            info!(
+                id = ?new_keypair.config().id(),
+                "bootstrapping: inserting key for new ciphersuite"
+            );
+            tx.put_global_hpke_keypair(&new_keypair).await?;
+
+            // If there are zero keypairs reported by the database, it's likely beacuse we're
+            // in a brand new database, and this is the first execution of Janus. Initialize
+            // the table with active keys. Janus won't be ready until there's at least one
+            // active keypair in the database.
+            //
+            // Otherwise, the ciphersuite was likely added to the configuration and its key
+            // needs to go through the normal pending->active state change.
+            if current_keypairs.is_empty() {
+                info!(id = ?new_keypair.config().id(), "bootstrapping: moving new key to active state");
+                tx.set_global_hpke_keypair_state(new_keypair.config().id(), &HpkeKeyState::Active)
+                    .await?;
+            }
+        } else {
+            let to_be_expired_keypairs: Vec<_> = active_keypairs
```

Comment:
I think I follow the logic here, but it's pretty complicated & procedural. I might just be missing an obvious policy, but it also seems somewhat ad-hoc. It is also challenging to be completely confident in this logic, as the logic switches between working with in-memory representations & actually touching the DB -- the in-memory representations will be out of sync with the state of the DB after we start touching the DB, so it's somewhat challenging to reason about this code.

I would suggest writing an explicit human-language policy for rotation, then implementing that policy. If there's no simple policy that matches the current code, a per-ciphersuite policy that might work would be (cribbing from [the `prio-server` key rotator](https://github.com/divviup/prio-server/blob/5aeb486a0f0a0d58a028cedcb9f6baa1f2bbac44/key-rotator/key/key.go#L214)):

1. For each pending key, if the pending key is at least age `active_min_age`, promote it to active.
1. If the ciphersuite is configured, and there are no pending/active keys or the newest pending/active key is at least age `creation_min_age`, create a new pending key.
1. While there are more than `maximum_active_keys` active keys, expire the oldest active key. [I think `maximum_active_keys` would typically be set to 1, and would be treated 0 for unconfigured ciphersuites. Maybe this doesn't need to be a config knob at all & can always be treated as 1 for configured ciphersuites, 0 for unconfigured?]
1. For each expired key, if the expired key is at least age `deletion_min_age`, delete it.

(It would be good to write this policy down in comments, near the code which implements each step. This policy is just an example that I think works, but maybe we can come up with a simpler one? Especially the special-casing around unconfigured ciphersuites might permit a simpler policy.)

To make the code clearer/more obviously correct, I suggest structuring it into three phases:
1) Read the existing DB state into an in-memory structure.
2) Manipulate the in-memory structure to perform rotation (in-memory). Ideally this would remove the "we have a bunch of in-memory stuff we make decisions on, but which is out of sync with changes we've already written to the DB" property the current code has, as decisions would be made at each step based on the current state of the in-memory structure, which would reflect the operations already performed on the in-memory structure.
3) Write the in-memory structure back to the DB, replacing the existing contents of the DB. This could be done by keeping track of a diff in the in-memory structure, generating a diff based on the in-memory structure, or even wiping the table & rewriting it with each rotation.

I would also like it if the handling for configured vs unconfigured ciphersuites was unified, as long as that doesn't introduce too much complexity. The suggested policy above does an OK job of this IMO, but maybe we can do better?

Reply:
> I would suggest writing an explicit human-language policy for rotation, then implementing that policy.

:+1:

> Read the existing DB state into an in-memory structure.
> ...
> Write the in-memory structure back to the DB, replacing the existing contents of the DB.

I had considered this early but chose not to since I felt the logic was simple enough, though borderline, that we could avoid the extra code.

In any case, I'm amenable to this idea. I was careful to not rely on fields that would have been outdated, but perhaps it's unnecessary mental overhead to have to consider the current table state versus in-memory state for each operation.

> I would also like it if the handling for configured vs unconfigured ciphersuites was unified, as long as that doesn't introduce too much complexity.

I disagree that we should handle unconfigured ciphersuites like we do configured ones. IMO if we encounter an unconfigured ciphersuite we should swiftly phase it out. This strikes me as the least surprising behavior.

That is, as an operator, the least surprising result of removing a ciphersuite from the configuration would be that the respective keys are removed from the system. We, as developers, balance that out by retaining the expired key for as long as we have to, but still immediately remove the key from advertisement.

The alternative approaches, i.e. continue rotating the key or let the key run out its lifespan, are surprising results, because now the operator has to go in and make manual changes to _actually_ remove the key from the system.

Or do you mean more mechanically, i.e. if we can sweep unconfigured keys in the same set of logic as configured keys?

Reply:
> I disagree that we should handle unconfigured ciphersuites like we do configured ones. IMO if we encounter an unconfigured ciphersuite we should swiftly phase it out. This strikes me as the least surprising behavior.
>
> That is, as an operator, the least surprising result of removing a ciphersuite from the configuration would be that the respective keys are removed from the system. We, as developers, balance that out by retaining the expired key for as long as we have to, but still immediately remove the key from advertisement.

OK -- I don't object to expiring keys for quickly, I think that can be done without making the policy too complicated. (N.B. the suggested policy would immediately expire any active keys for an unconfigured policy.)

I do think we want to let expired keys for unconfigured ciphersuites time out before deletion, but I think we agree on that point.

> Or do you mean more mechanically, i.e. if we can sweep unconfigured keys in the same set of logic as configured keys?

Yes, this would be nice. The current logic has many special cases; while we do need (effectively) a `ciphersuite_configured` signal, I think it should basically feed into the same set of steps used by configured ciphersuites. Basically I want to avoid having two different codepaths for configured vs unconfigured ciphersuites.